File: solutions/python/palindrome-products/1/palindrome_products.py
def is_palindrome(n):
    """Check if a number is a palindrome."""
    s = str(n)
    return s == s[::-1]


def smallest(min_factor, max_factor):
    """Find the smallest palindromic product of factors in the given range."""
    if min_factor > max_factor:
        raise ValueError("min must be <= max")
    
    palindromes = {}  # value -> set of factor pairs
    
    for i in range(min_factor, max_factor + 1):
        for j in range(i, max_factor + 1):
            product = i * j
            if is_palindrome(product):
                if product not in palindromes:
                    palindromes[product] = set()
                palindromes[product].add(tuple(sorted([i, j])))
    
    if not palindromes:
        return None, []
    
    min_val = min(palindromes.keys())
    factors = sorted(list(palindromes[min_val]))
    return min_val, factors


def largest(min_factor, max_factor):
    """Find the largest palindromic product of factors in the given range."""
    if min_factor > max_factor:
        raise ValueError("min must be <= max")
    
    palindromes = {}  # value -> set of factor pairs
    
    for i in range(min_factor, max_factor + 1):
        for j in range(i, max_factor + 1):
            product = i * j
            if is_palindrome(product):
                if product not in palindromes:
                    palindromes[product] = set()
                palindromes[product].add(tuple(sorted([i, j])))
    
    if not palindromes:
        return None, []
    
    max_val = max(palindromes.keys())
    factors = sorted(list(palindromes[max_val]))
    return max_val, factors

File: solutions/ruby/palindrome-products/1/palindrome_products.rb
class Palindromes
  attr_reader :smallest, :largest

  def initialize(max_factor:, min_factor: 1)
    @max_factor = max_factor
    @min_factor = min_factor
    raise ArgumentError, "min must be <= max" if min_factor > max_factor
  end

  def generate
    palindromes = Hash.new { |h, k| h[k] = [] }
    (@min_factor..@max_factor).each do |a|
      (a..@max_factor).each do |b|
        product = a * b
        if palindrome?(product)
          palindromes[product] << [a, b]
        end
      end
    end
    smallest_val = palindromes.keys.min
    largest_val = palindromes.keys.max
    @smallest = Palindrome.new(smallest_val, palindromes[smallest_val] || [])
    @largest = Palindrome.new(largest_val, palindromes[largest_val] || [])
  end

  def palindrome?(n)
    s = n.to_s
    s == s.reverse
  end
end

class Palindrome
  attr_reader :value, :factors

  def initialize(value, factors)
    @value = value
    @factors = factors
  end
end
